### PR TITLE
Add calculating long wav frr again

### PR DESCRIPTION
**Pull request by blessyyyu**

### CODE REVIEW

File: kws/bin/score_longwav.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+                # mask对应的true的部分用0填充
+                # Getting every frames desn't need to mask
+                # logits = logits.masked_fill(mask, 0.0)
+                logits = logits.cpu()
+                for i in range(len(keys)):
+                    key = keys[i]
+                    score = logits[i][:lengths[i]]
+                    score = score[:, int(keyword_label)]
+                    # keep 2 significant digits
+                    score = ' '.join([str("%.2g" % x) for x in score.tolist()])
```
Comment (robin1001) [inline kws/bin/score_longwav.py:140]: python str '{:.2}'.format(x);

Comment (robin1001) [reply kws/bin/score_longwav.py:140]: score = ' '.join(['{.3f}'.format(x) for x in ...]

File: kws/bin/compute_det_longwav.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import json
+
+
+def load_label_and_score(keyword, label_file, score_file):
+    # utt_id : score list
```
Comment (robin1001) [inline kws/bin/compute_det_longwav.py:21]: remove useless comment

File: kws/bin/score_longwav.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    for abs_path in score_file_list:
+        with torch.no_grad(), open(abs_path, 'w', encoding='utf8') as fout:
+            keyword_label = abs_path.split('/')[-1].split('.')[1]
+            # print('keyword_label = ', keyword_label)
+            for batch_idx, batch in enumerate(test_data_loader):
+                keys, feats, target, lengths = batch
+                feats = feats.to(device)
+                lengths = lengths.to(device)
+                # mask = padding_mask(lengths).unsqueeze(2)
```
Comment (robin1001) [inline kws/bin/score_longwav.py:129]: remove it, it's useless

File: kws/bin/score_longwav.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                        help='num of subprocess workers for reading')
+    parser.add_argument('--pin_memory',
+                        action='store_true',
+                        default=False,
+                        help='Use pinned memory buffers used for reading')
+    parser.add_argument('--prefetch',
+                        default=100,
+                        type=int,
+                        help='prefetch number')
+    parser.add_argument('--score_file_dir',
```
Comment (robin1001) [inline kws/bin/score_longwav.py:58]: write score in just one file

File: kws/bin/compute_det_longwav.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    parser.add_argument('--test_data', required=True, help='label file')
+    parser.add_argument('--keyword', type=int, default=0, help='score file')
+    parser.add_argument('--score_file', required=True, help='score file')
+    parser.add_argument('--step', type=float, default=0.01, help='score file')
+    parser.add_argument('--stats_file',
+                        required=True,
+                        help='false reject/alarm stats file')
+    args = parser.parse_args()
+    # 'window_shift' is used to skip the frames after triggered
+    window_shift = 50
```
Comment (robin1001) [inline kws/bin/compute_det_longwav.py:68]: set is as a parameter

**Review by robin1001 — CHANGES_REQUESTED** [review]
please see inline.

File: examples/hi_xiaowen/s0/run.sh [hunk 6]
```diff
@@ -95,19 +95,22 @@ if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
...
   mkdir -p $result_dir
-  python kws/bin/score.py \
+  python kws/bin/score_longwav.py \
     --config $dir/config.yaml \
     --test_data data/test/data.list \
     --batch_size 256 \
     --checkpoint $score_checkpoint \
-    --score_file $result_dir/score.txt \
+    --score_file $result_dir/score_longwav.txt  \
+    --num_keywords $num_keywords  \
```
Comment (robin1001) [inline examples/hi_xiaowen/s0/run.sh:104]: this parameter is not required

**Review by robin1001 — CHANGES_REQUESTED** [review]
please see inline.

File: kws/bin/compute_det.py [hunk 7]
```diff
@@ -14,16 +15,17 @@
...
+    score_table = defaultdict(list)
     with open(score_file, 'r', encoding='utf8') as fin:
         for line in fin:
             arr = line.strip().split()
             key = arr[0]
-            score = float(arr[keyword + 1])
-            score_table[key] = score
+            str_list = arr[1:]
+            scores = list(map(float, str_list))
+            score_table[key].append(scores)
```
Comment (robin1001) [inline kws/bin/compute_det.py:28]: 1. score_table.get(key, [])
2. if key not in score_table: score_table[key] = []

File: kws/bin/compute_det.py [hunk 8]
```diff
@@ -44,36 +46,49 @@ def load_label_and_score(keyword, label_file, score_file):
...
                         help='false reject/alarm stats file')
     args = parser.parse_args()
-
+    window_shift = args.window_shift
     keyword_table, filler_table, filler_duration = load_label_and_score(
         args.keyword, args.test_data, args.score_file)
     print('Filler total duration Hours: {}'.format(filler_duration / 3600.0))
-
     with open(args.stats_file, 'w', encoding='utf8') as fout:
+        keyword_index = int(args.stats_file.split('/')[-1].split('.')[1])
```
Comment (robin1001) [inline kws/bin/compute_det.py:66]: use args.keyword

File: kws/bin/compute_det.py [hunk 9]
```diff
@@ -14,16 +15,17 @@
...
 import argparse
 import json
-
+from collections import defaultdict
 
 def load_label_and_score(keyword, label_file, score_file):
-    score_table = {}
+    score_table = defaultdict(list)
     with open(score_file, 'r', encoding='utf8') as fin:
         for line in fin:
```
Comment (robin1001) [inline kws/bin/compute_det.py:23]:
``` python
for i, line in enumerate(fin):
    if i % num_keyword == keyword:
 ```

**Review by robin1001 — CHANGES_REQUESTED** [review]
please see inline.

**Review by robin1001 — APPROVED** [review]

**Comment by blessyyyu** [comment]
Okay, I'm extremely grateful for your efforts on reviewing and instruct, and I will finish it on tomorrow morning. Good night every one.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Binbin ***@***.***&gt;; 
发送时间: 2022年3月23日(星期三) 晚上11:15
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [wenet-e2e/wekws] Add calculating long wav frr again (PR #64)





 
@blessyyyu please override score.py and compute_det.py with the new long implementation.

**Pull request merged**